packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * WordPress dependencies
+ */
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+import { decodeEntities } from '@wordpress/html-entities';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Filters the search by type
+ *
+ * @typedef { 'post' | 'term' | 'post-format' } WPLinkSearchType
+ */
+
+/**
+ * @typedef WPLinkSearchOptions
+ *
+ * @property {boolean}             [isInitialSuggestions] Displays initial search suggestions, when true.
+ * @property {WPLinkSearchType}    [type]                 Filters by search type.
+ * @property {string}              [subtype]              Slug of the post-type or taxonomy.
+ * @property {number}              [page]                 Which page of results to return.
+ * @property {number}              [perPageArg]           Search results per page.
+ */
+
+/**
+ * @typedef WPLinkSearchResult
+ *
+ * @property {number} id     Post or term id.
+ * @property {string} url    Link url.
+ * @property {string} title  Title of the link.
+ * @property {string} type   The taxonomy or post type slug or type URL.
+ */
+
+/**
+ * Fetches link suggestions from the API.
+ *
+ * @async
+ * @typedef {( searchText: string, searchOptions: WPLinkSearchOptions ) => Promise< WPLinkSearchResult[] >} WPFetchLinkSuggestions
+ */
+
+/**
+ * Returns a function that when invoked, fetches link suggestions from the API.
+ *
+ * @param {Object}   [editorSettings]
+ * @param {boolean}  [editorSettings.disablePostFormats]
+ *
+ * @example
+ * ```js
+ * import { __experimentalFetchLinkSuggestions as createFetchLinkSuggestions } from '@wordpress/core-data';
+ *
+ * //...
+ *
+ * export function initialize( id, settings ) {
+ *
+ * settings.__experimentalFetchLinkSuggestions = createFetchLinkSuggestions(
+		settings
+	);
+ * ```
+ * @return { WPFetchLinkSuggestions } Function that fetches link suggestions
+ *
+ */
+const createFetchLinkSuggestions = ( {
+	disablePostFormats = false,
+} = {} ) => async ( search, searchOptions ) => {
+	const {
+		isInitialSuggestions = false,
+		type = undefined,
+		subtype = undefined,
+		page = undefined,
+		perPageArg = undefined,
+	} = searchOptions;
+	const perPage = perPageArg || isInitialSuggestions ? 3 : 20;
+
+	const queries = [];
+
+	if ( ! type || type === 'post' ) {
+		queries.push(
+			apiFetch( {
+				path: addQueryArgs( '/wp/v2/search', {
+					search,
+					page,
+					per_page: perPage,
+					type: 'post',
+					subtype,
+				} ),
+			} ).catch( () => [] ) // fail by returning no results
+		);
+	}
+
+	if ( ! type || type === 'term' ) {
+		queries.push(
+			apiFetch( {
+				path: addQueryArgs( '/wp/v2/search', {
+					search,
+					page,
+					per_page: perPage,
+					type: 'term',
+					subtype,
+				} ),
+			} ).catch( () => [] )
+		);
+	}
+
+	if ( ! disablePostFormats && ( ! type || type === 'post-format' ) ) {
```

Comment:
It might be worth taking this opportunity to write some unit tests for this function since  it has a bunch of conditional logic that is often error prone and `apiFetch` is easy to mock.